lms/djangoapps/grades/tests/test_tasks.py

```diff
@@ -0,0 +1,117 @@
+"""
+Tests for the functionality and infrastructure of grades tasks.
+"""
+
+import ddt
+from django.conf import settings
+from mock import patch
+from unittest import skip
+
+from student.models import anonymous_id_for_user
+from student.tests.factories import UserFactory
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory, ItemFactory, check_mongo_calls
+
+from lms.djangoapps.grades.config.models import PersistentGradesEnabledFlag
+from lms.djangoapps.grades.tasks import recalculate_subsection_grade
+
+
+@patch.dict(settings.FEATURES, {'PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS': False})
+@ddt.ddt
+class RecalculateSubsectionGradeTest(ModuleStoreTestCase):
+    """
+    Ensures that the recalculate subsection grade task functions as expected when run.
+    """
+    def setUp(self):
+        super(RecalculateSubsectionGradeTest, self).setUp()
+        self.user = UserFactory()
+        PersistentGradesEnabledFlag.objects.create(enabled_for_all_courses=True, enabled=True)
+
+    def set_up_course(self, enable_subsection_grades=True):
+        """
+        Configures the course for this test.
+        """
+        # pylint: disable=attribute-defined-outside-init,no-member
+        self.course = CourseFactory.create(
+            org='edx',
+            name='course',
+            run='run',
+        )
+        if not enable_subsection_grades:
+            PersistentGradesEnabledFlag.objects.create(enabled=False)
+
+        self.chapter = ItemFactory.create(parent=self.course, category="chapter", display_name="Chapter")
+        self.sequential = ItemFactory.create(parent=self.chapter, category='sequential', display_name="Open Sequential")
+        self.problem = ItemFactory.create(parent=self.sequential, category='problem', display_name='problem')
+
+        self.score_changed_kwargs = {
+            'points_possible': 10,
+            'points_earned': 5,
+            'user': self.user.id,
+            'course_id': unicode(self.course.id),
+            'usage_id': unicode(self.problem.location),
+        }
+
+        # this call caches the anonymous id on the user object, saving 4 queries in all happy path tests
+        _ = anonymous_id_for_user(self.user, self.course.id)
+        # pylint: enable=attribute-defined-outside-init,no-member
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_subsection_grade_updated_on_signal(self, default_store):
+        with self.store.default_store(default_store):
+            self.set_up_course()
+            self.assertTrue(PersistentGradesEnabledFlag.feature_enabled(self.course.id))
+            with check_mongo_calls(2) and self.assertNumQueries(13):
```

Comment:
It's because I'm now sending `user.id` to the task as input, and translating that value into a `User` object via `student = User.objects.get(id=kwargs['user'])`. That hits the db twice (`SELECT FROM "auth_user"` and `SELECT FROM "student_courseaccessrole"`).

This change was made when going to a task, as all inputs need to be json-serializable and the `User` object we had been using before is not. Any suggestions for better ways to handle this?